<a href="https://colab.research.google.com/github/darome-ur/SistemasDP1/blob/master/NoteBook3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NoteBook 3

A continuación vamos a realizar una comparativa de distintos algoritmos de clasificación para el dataset proporcionado. Cargamos los datos y las librerias.

In [0]:
import pandas as pd
from sklearn.model_selection import train_test_split
df = pd.read_csv('train.csv')
X =  df.values[:,:-1]
Y =  df.values[:,-1]

In [0]:
(trainData, testData, trainLabels, testLabels) = train_test_split(X,Y,test_size=0.25, random_state=42)

In [0]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import LogisticRegression
'''para evitar los warnings'''
import warnings; warnings.simplefilter('ignore')

Vamos a empezar por el primer algoritmo el KNN.

In [5]:
modelKNN = KNeighborsClassifier(n_neighbors=5)
modelKNN.fit(trainData, trainLabels)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [6]:
Clases=modelKNN.predict(testData)
modelKNN.predict(testData)

array([3., 3., 3., ..., 3., 3., 4.])

Vamos a seguir con el algoritmo del Arbol de decisión.

In [7]:
modelArbol = DecisionTreeClassifier(random_state=84)
modelArbol.fit(trainData, trainLabels)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=84, splitter='best')

In [8]:
modelArbol.predict(testData)

array([3., 3., 2., ..., 3., 3., 4.])

Vamos a seguir con el algoritmo del Regresion Logistica.

In [0]:
modelLR = LogisticRegression(random_state=84)

Vamos a seguir con el algoritmo SVM.

In [10]:
'''modelSVMLineal = SVC(kernel="linear")
modelSVMLineal.fit(trainData, trainLabels)
modelSVMLineal.predict(testData)'''


'modelSVMLineal = SVC(kernel="linear")\nmodelSVMLineal.fit(trainData, trainLabels)\nmodelSVMLineal.predict(testData)'

Es una burrada lo que cuesta el entrenamiento, lo he cancelado y decidido prescindir de él.

Vamos a ver el algoritmo de Redes neuronales

In [11]:
modelPerceptron=MLPClassifier(momentum= 0.95, learning_rate_init= 0.1, hidden_layer_sizes= (5, 3, 2), activation= 'identity')
modelPerceptron.fit(trainData, trainLabels)

MLPClassifier(activation='identity', alpha=0.0001, batch_size='auto',
              beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(5, 3, 2), learning_rate='constant',
              learning_rate_init=0.1, max_fun=15000, max_iter=200,
              momentum=0.95, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=None, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

Este sera la funcion que utilizaremos para evaluar. Es el metodo Gridseach.

In [0]:
from sklearn.model_selection import GridSearchCV
scores = ['precision', 'recall']

In [0]:
def gridsearch(model,parameters,scores,k):
    for score in scores:
        print("# Tuning hyper-parameters for %s" % score)
        print()

        clf = GridSearchCV(model, parameters, cv=k,
                           scoring='%s_macro' % score)
        clf.fit(trainData, trainLabels)

        print("Best parameters set found on development set:")
        print()
        print(clf.best_params_)
        print()
        print("Grid scores on development set:")
        print()
        means = clf.cv_results_['mean_test_score']
        stds = clf.cv_results_['std_test_score']
        for mean, std, params in zip(means, stds, clf.cv_results_['params']):
            print("%0.3f (+/-%0.03f) for %r"
                  % (mean, std * 2, params))
        print()

In [14]:

KNN_parameters =  [{'n_neighbors': range(1,51),
                     'metric': ["euclidean","manhattan","chebyshev","minkowski"]}]
%time
gridsearch(modelKNN,KNN_parameters,scores,5)

CPU times: user 3 µs, sys: 1 µs, total: 4 µs
Wall time: 6.91 µs
# Tuning hyper-parameters for precision

Best parameters set found on development set:

{'metric': 'manhattan', 'n_neighbors': 7}

Grid scores on development set:

0.579 (+/-0.017) for {'metric': 'euclidean', 'n_neighbors': 1}
0.589 (+/-0.012) for {'metric': 'euclidean', 'n_neighbors': 2}
0.586 (+/-0.020) for {'metric': 'euclidean', 'n_neighbors': 3}
0.588 (+/-0.017) for {'metric': 'euclidean', 'n_neighbors': 4}
0.595 (+/-0.021) for {'metric': 'euclidean', 'n_neighbors': 5}
0.585 (+/-0.021) for {'metric': 'euclidean', 'n_neighbors': 6}
0.593 (+/-0.027) for {'metric': 'euclidean', 'n_neighbors': 7}
0.591 (+/-0.032) for {'metric': 'euclidean', 'n_neighbors': 8}
0.587 (+/-0.030) for {'metric': 'euclidean', 'n_neighbors': 9}
0.583 (+/-0.025) for {'metric': 'euclidean', 'n_neighbors': 10}
0.588 (+/-0.019) for {'metric': 'euclidean', 'n_neighbors': 11}
0.587 (+/-0.026) for {'metric': 'euclidean', 'n_neighbors': 12}
0.586 (+/-0.0

In [15]:

arbol_parameters = [{'criterion': ["gini","entropy"],
                     'max_features': [5],
                     'min_samples_split': range(2,21)}]
%time
gridsearch(modelArbol,arbol_parameters,scores,5)                    

CPU times: user 3 µs, sys: 0 ns, total: 3 µs
Wall time: 7.15 µs
# Tuning hyper-parameters for precision

Best parameters set found on development set:

{'criterion': 'gini', 'max_features': 5, 'min_samples_split': 20}

Grid scores on development set:

0.648 (+/-0.025) for {'criterion': 'gini', 'max_features': 5, 'min_samples_split': 2}
0.651 (+/-0.025) for {'criterion': 'gini', 'max_features': 5, 'min_samples_split': 3}
0.650 (+/-0.024) for {'criterion': 'gini', 'max_features': 5, 'min_samples_split': 4}
0.657 (+/-0.020) for {'criterion': 'gini', 'max_features': 5, 'min_samples_split': 5}
0.653 (+/-0.020) for {'criterion': 'gini', 'max_features': 5, 'min_samples_split': 6}
0.661 (+/-0.024) for {'criterion': 'gini', 'max_features': 5, 'min_samples_split': 7}
0.664 (+/-0.024) for {'criterion': 'gini', 'max_features': 5, 'min_samples_split': 8}
0.664 (+/-0.024) for {'criterion': 'gini', 'max_features': 5, 'min_samples_split': 9}
0.670 (+/-0.026) for {'criterion': 'gini', 'max_features': 5

In [16]:

perceptron_parameters = [{'hidden_layer_sizes':[(5,2),(3,3,3),(5,3,2),(5,4,3,2)],
                               'activation':['identity','logistic','tanh','relu'],
                               'learning_rate_init':[0.001,0.01,0.1],
                               'momentum':[0.9,0.95,0.99]
                               }]
%time
'''gridsearch(modelPerceptron,perceptron_parameters,scores,5)'''                          

CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 7.87 µs


'gridsearch(modelPerceptron,perceptron_parameters,scores,5)'

Es inviable lo que llega a tardar.

In [17]:
LR_parameters = [{'C': [0.1,0.5,1, 10, 100, 1000]}]
%time
gridsearch(modelLR,LR_parameters,scores,5)

CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 7.39 µs
# Tuning hyper-parameters for precision

Best parameters set found on development set:

{'C': 0.5}

Grid scores on development set:

0.191 (+/-0.054) for {'C': 0.1}
0.192 (+/-0.054) for {'C': 0.5}
0.191 (+/-0.053) for {'C': 1}
0.188 (+/-0.054) for {'C': 10}
0.189 (+/-0.054) for {'C': 100}
0.190 (+/-0.054) for {'C': 1000}

# Tuning hyper-parameters for recall

Best parameters set found on development set:

{'C': 1000}

Grid scores on development set:

0.205 (+/-0.025) for {'C': 0.1}
0.204 (+/-0.026) for {'C': 0.5}
0.203 (+/-0.024) for {'C': 1}
0.203 (+/-0.030) for {'C': 10}
0.204 (+/-0.026) for {'C': 100}
0.207 (+/-0.024) for {'C': 1000}



Vemos que de todos los algoritmos, los que mejor precision sacan son: KNN, y árboles.

In [0]:
from sklearn.metrics import classification_report

In [21]:
print("EVALUACIÓN EN CONJUNTO DE TEST USANDO KNN")
predictionsKNN = modelKNN.predict(testData)
print(classification_report(testLabels, predictionsKNN))

EVALUACIÓN EN CONJUNTO DE TEST USANDO KNN
              precision    recall  f1-score   support

         0.0       0.77      0.70      0.74       875
         1.0       0.55      0.68      0.61        95
         2.0       0.55      0.63      0.59       433
         3.0       0.51      0.61      0.56       624
         4.0       0.60      0.53      0.56       662
         5.0       0.50      0.42      0.45       311

    accuracy                           0.60      3000
   macro avg       0.58      0.60      0.58      3000
weighted avg       0.61      0.60      0.61      3000



In [23]:
print("EVALUACIÓN EN CONJUNTO DE TEST USANDO AEBOLES")
predictionsArboles = modelArbol.predict(testData)
print(classification_report(testLabels, predictionsArboles))

EVALUACIÓN EN CONJUNTO DE TEST USANDO AEBOLES
              precision    recall  f1-score   support

         0.0       0.93      0.82      0.87       875
         1.0       0.72      0.76      0.74        95
         2.0       0.62      0.66      0.64       433
         3.0       0.58      0.62      0.60       624
         4.0       0.57      0.57      0.57       662
         5.0       0.51      0.56      0.53       311

    accuracy                           0.67      3000
   macro avg       0.65      0.66      0.66      3000
weighted avg       0.68      0.67      0.67      3000



**Conclusión**: el algoritmo de arbol es el que mas precisión da. Arbol un 67% (+/-2.6%) y KNN 61.3% (+/-1.5%)

A continuación vamos a ver si con una estandarización o normalizacion mejoran: mejoran:

In [0]:
from sklearn.preprocessing import StandardScaler
stdsc = StandardScaler()
X_train_std = stdsc.fit_transform(X_train)
X_test_std = stdsc.transform(X_test)

In [0]:
from sklearn.preprocessing import MinMaxScaler
mms = MinMaxScaler()
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.25, random_state=0)
X_train_norm = mms.fit_transform(X_train)
X_test_norm = mms.transform(X_test)

In [0]:
X_train_norm = mms.fit_transform(X_train)
X_test_norm = mms.transform(X_test)

In [30]:
modelKNN.fit(X_train_norm,Y_train)
print('Training accuracy:', modelKNN.score(X_train_norm,Y_train))
print('Test accuracy:', modelKNN.score(X_test_norm,Y_test))

Training accuracy: 0.7391932436937437
Test accuracy: 0.6526666666666666


In [32]:
modelKNN.fit(X_train_std,Y_train)
print('Training accuracy:', modelKNN.score(X_train_std,Y_train))
print('Test accuracy:', modelKNN.score(X_test_std,Y_test))

Training accuracy: 0.7356372930325592
Test accuracy: 0.643


In [33]:
modelArbol.fit(X_train_norm,Y_train)
print('Training accuracy:', modelArbol.score(X_train_norm,Y_train))
print('Test accuracy:', modelArbol.score(X_test_norm,Y_test))

Training accuracy: 0.999888876541838
Test accuracy: 0.68


In [34]:
modelArbol.fit(X_train_std,Y_train)
print('Training accuracy:', modelArbol.score(X_train_std,Y_train))
print('Test accuracy:', modelArbol.score(X_test_std,Y_test))

Training accuracy: 1.0
Test accuracy: 0.681


**Conclusion:** vemos que el arbol sube la precisión al 68.1% frente al 67% +-2.6% y con KNN normalizando se mejora a 65.266% frente a un 61.3% (+/-1.5%)